lib/PhasarLLVM/TypeHierarchy/DIBasedTypeHierarchy.cpp

```diff
@@ -0,0 +1,272 @@
+          if (TransitiveClosure[CompareRow][Column] &&
+              !TransitiveClosure[CurrentRow][Column] && CurrentRow != Column) {
+            TransitiveClosure[CurrentRow][Column] = true;
+            Change = true;
+          }
+        }
+      }
+    }
+  }
+
+  // add edges onto vertices themselves
+  for (size_t I = 0; I < TransitiveClosure.size(); I++) {
+    TransitiveClosure[I][I] = true;
+  }
+
+  // get VTables
+  const auto Subprograms = Finder.subprograms();
+  for (auto *Subprogram : Subprograms) {
+
+    if (const auto *SubProgramType =
+            llvm::dyn_cast<llvm::DISubprogram>(Subprogram)) {
+      if (!SubProgramType->getVirtuality()) {
+        continue;
+      }
+      // get all virtual functions
+      std::vector<const llvm::Function *> VirtualFunctions;
+      for (const auto &Function : Module->functions()) {
+        if (SubProgramType->getLinkageName() ==
+            Function.stripPointerCasts()->getName().str()) {
+          VirtualFunctions.push_back(&Function);
+        }
+      }
+      VTables.push_back(LLVMVFTable(VirtualFunctions));
+    }
+  }
+}
+
+[[nodiscard]] bool DIBasedTypeHierarchy::isSubType(ClassType Type,
+                                                   ClassType SubType) {
+  // find index of super type
+  unsigned long IndexOfType = TypeToVertex.find(Type)->getSecond();
+  unsigned long IndexOfSubType = TypeToVertex.find(SubType)->getSecond();
+
+  // if the super type or the sub type haven't been found, return false
+  if (IndexOfType >= TypeToVertex.size() ||
+      IndexOfSubType >= TypeToVertex.size()) {
+    return false;
+  }
+
+  return TransitiveClosure[IndexOfType][IndexOfSubType];
+}
+
+[[nodiscard]] auto DIBasedTypeHierarchy::getSubTypes(ClassType Type)
+    -> std::set<ClassType> {
+  // find index of super type
+  unsigned long IndexOfType = TypeToVertex.find(Type)->getSecond();
+
+  // if the super type hasn't been found, return an empty set
+  if (IndexOfType >= TypeToVertex.size()) {
+    return {};
+  }
+
+  // return all sub types
+  std::set<ClassType> SubTypes = {};
+
+  size_t Index = 0;
+  for (size_t I = 0; I < TransitiveClosure[IndexOfType].size(); I++) {
+    if (TransitiveClosure[IndexOfType][I]) {
+      SubTypes.insert(VertexTypes[Index]);
+    }
+
+    Index++;
+  }
+
+  return SubTypes;
+}
+
+[[nodiscard]] bool DIBasedTypeHierarchy::isSuperType(ClassType Type,
+                                                     ClassType SuperType) {
+  return isSubType(SuperType, Type); // NOLINT
+}
+
+[[nodiscard]] auto DIBasedTypeHierarchy::getSuperTypes(ClassType Type)
+    -> std::set<ClassType> {
+  // TODO: implement (low priority)
+  llvm::report_fatal_error("Not implemented");
+}
+
+[[nodiscard]] bool DIBasedTypeHierarchy::hasVFTable(ClassType Type) const {
+  return !VTables.empty();
+}
+
+[[nodiscard]] auto DIBasedTypeHierarchy::getVFTable(ClassType Type) const
+    -> const VFTable<f_t> * {
+
+  return &(VTables.at(TypeToVertex.find(Type)->getSecond()));
```

Comment:
TODO: Error handling for Type not found in TypeToVertex + assert inbounds + replace at() by []